[WIP] Mwilck/multipath 15.4

mwilck opened the pull request:
### Update multipath chapter for 15.4

The multipath chapter is strongly outdated

### PR creator: Are there any relevant issues/feature requests?

* bsc#1179182
* jsc#SLE-19638


### PR creator: Which product versions do the changes apply to?

When opening a PR, check all versions of the documentation that your PR applies to.

- SLE 15/openSUSE Leap 15.x
  - [x ] SLE 15 SP4/openSUSE Leap 15.4 *(current `main`, no backport necessary)*
  - [ x] SLE 15 SP3/openSUSE Leap 15.3
  - [ x] SLE 15 SP2/openSUSE Leap 15.2
  - [ x] SLE 15 SP1
  - [ x] SLE 15 GA
- SLE 12
  - [x ] SLE 12 SP5
  - [x] SLE 12 SP4
  - [x] SLE 12 SP3

### PR reviewer only: Have all backports been applied?

The doc team member merging your PR will take care of backporting to older documents.
When opening a PR, do *not* set the following check box.

- [ ] all necessary backports are done

*Please do not merge: WIP, for now just to make the changes easily review-able*

cjschroder commented:
Thank you @mwilck!

mwilck commented:
I'm not done yet, unfortunately. As you can see from the many fixups, it's still work in progress.

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
+      is done in user space by the
+      <command>kpartx</command> tool. <command>kpartx</command> is
+      automatically invoked by udev rules; there is usually no need to run it
+      manually. Technically, &ldquo;partition&rdquo; devices created by kpartx are
```
```suggestion
      manually. Technically, &ldquo;partition&rdquo; devices created by <command>kpartx</command> are
```

lvicoun replied:
Is there any term that can be used without the quotes?

mwilck replied:
None that would describe these devices better.

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
+      <filename>/dev/disk/by-id/dm-uuid-part<replaceable>N</replaceable>-mpath-<replaceable>WWID</replaceable></filename>.
+    </para>
+    <note>
+      <title>Disabling invocation of kpartx</title>
```
```suggestion
      <title>Disabling invocation of <command>kpartx</command></title>
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
+      <filename>/etc/multipath/prkeys</filename> are maintained by
+      <systemitem>multipath-tools</systemitem> to store persistent information
+      about previously created multipath maps, map names, and reservation keys
+      for SCSI persistent reservations, respectively. They are
```
```suggestion
      for SCSI persistent reservations, respectively. 
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
+      <systemitem>multipath-tools</systemitem> to store persistent information
+      about previously created multipath maps, map names, and reservation keys
+      for SCSI persistent reservations, respectively. They are
+      &ldquo;configuration files&rdquo; in a broader sense, but not meant to
```
```suggestion
      Do not edit these generated configuration files.
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
+      about previously created multipath maps, map names, and reservation keys
+      for SCSI persistent reservations, respectively. They are
+      &ldquo;configuration files&rdquo; in a broader sense, but not meant to
+      be manually edited.
```
```suggestion
      
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
+    <para>
+      Except for <filename>/etc/multipath.conf</filename>, the paths of
+      the configuration directories and files are configurable,
+      but this is discouraged. It will be disallowed in future versions
```
I'd omit the information now. Now it is possible to configure the paths. If it is not recommended due to another reason, please state the reason. Changes for future releases are a matter of next releases documentation :)

mwilck replied:
I would like to mention that this is discouraged.

mwilck replied:
I removed the part about the future change.

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
-    the multipath maps as described in
-    <xref
-    linkend="sec-multipath-conf-file-apply"/>.
+     It is possible to generate a <filename>multipath.conf</filename> template
```
```suggestion
     You can generate a <filename>multipath.conf</filename> template
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
+    <title>multipath.conf Syntax</title>
+    <para>
+      The <filename>/etc/multipath.conf</filename> file uses a hierarchy
+      of sections, subsections, and attribute / value pairs.
```
```suggestion
      of sections, subsections, and attribute/value pairs.
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
+      </listitem>
+      <listitem>
+	<para>
+	  Attributes and section names must be keywords. The allowd keywords are
```
```suggestion
	  Attributes and section names must be keywords. The allowed keywords are
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
-       that do not support individual settings, these values overwrite what is
-       specified in the <literal>defaults</literal> and
-       <literal>devices</literal> sections of the configuration file.
+	Storage-controller-specific settings. This section is
```
```suggestion
	Settings specific to the storage controller. This section is
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
-    effect when <command>multipathd</command> is running. After you make
-    changes, save and close the file, then run the following commands to apply
-    the changes and update the multipath maps:
+     Don't forget to synchronize with the configuration in the inird, see
```
```suggestion
     Don't forget to synchronize with the configuration in the initrd, see
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
+     <screen>multipath -d -v2</screen>
+     <para>
+       This command shows new maps to be created with the proposed topology.
+       It doesn't show whether maps will be removed / flushed, though. To obtain
```
```suggestion
       However, the command doesn't show whether maps will be removed/flushed. To obtain
```

lvicoun reviewed (commented):
Martin, thanks a lot for the review! You've done a great job!

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -37,201 +35,246 @@
+      configurable and flexible.
+    </para>
+    <para>
+      The Linux <emphasis>NVM Express</emphasis> (NVMe) kernel subsystem implements
```
It's more than that. You can call it a collection of kernel modules, but "subsystem" seems to be the correct term IMO.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -37,201 +35,246 @@
+	    active/passive (failover) and active/active
+	    (load-balancing) storage array configurations were
+	    distinguished. These concepts still exist, but they are merely
+	    special cases of the concepts of path groups and asymmetric access
```
yes

mwilck replied:
ALUA = "Asymmetric Logical Unit Access"

mwilck replied:
I actually explained this in the hardware handlers section below.

I'm removing the word "asymmetric" here.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -284,51 +321,144 @@
+	  <para>
+	    Enabling multipathing in the initramfs, but configuring
+	    <systemitem>multipath-tools</systemitem> to ignore the local disk.
+	    This can be achieved after installation by blacklisting, or with the
```
Yes. I haven't finished that part yet.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -37,201 +35,246 @@
-      file, you must run <command>dracut</command> <option>-f</option> to
-      re-create the <systemitem>initrd</systemitem> on your system. Reboot
-      afterward for the changes to take effect.
+       The fact that multipath-tools has built-in presets for a given
```
This would be misleading, IMO. multipath-tools  **has**  presets built in for a lot of storage arrays. I'll try to find a better wording.

mwilck replied:
I've reworked this paragraph.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -37,201 +35,246 @@
-      re-create the <systemitem>initrd</systemitem> on your system. Reboot
-      afterward for the changes to take effect.
+       The fact that multipath-tools has built-in presets for a given
+       storage array does not imply that this specific array has been tested
```
No, we need to keep "does not imply". The point here is that users / customers can't infer any sort of support or guarantee from the fact that multipath-tools has some presets built in for their array. "might not" is too weak / imprecise IMO.

mwilck replied:
I've reworked this.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -284,51 +321,144 @@
+	  This configuration results if multipathing was enabled during
+	  initial system installation with &yast;.
+	  It is the simplest configuration, but has a performance
+	  overhead.
```
I don't know. Nobody has ever benchmarked this systematically. It would be very easy to make false claims here. I'd rather be vague for the time being.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -284,51 +321,144 @@
+      <title>Keeping the initial RAM disk synchronized</title>
+      <important>
+	<para>
+	  Make sure
```
I don't understand your question

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -284,51 +321,144 @@
+	  changes.
+	</para>
+      </important>
+      <para>If multipathing is enabled in the system, it also needs to be
```
yes. I believe it's pretty much common knowledge, but some people may need guidance.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -355,24 +485,15 @@
-    system will not properly boot and the start-up procedure will result in an
-    emergency shell. See <xref linkend="sec-multipath-trouble-root"/> for
-    instructions on how to avoid or repair such a scenario.
+     Special care must be taken when using multipathing together with
```
I'm no clustering expert. I an change the workding from "is often disabled" to "should be disabled", but clustering w/multipath is a highly complex setup that can't be treated comprehensively in generic documentation.

mwilck replied:
Btw we also have a chapter about clustering. Perhaps the authors of that chapter should review this part and discuss it with me.

mwilck replied:
I've shortened the paragraph and added an xref to the "configuring policies" section instead, where this is explained in more detail.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -355,24 +485,15 @@
+     cluster solutions (<option>no_path_retry fail</option>), because it might
+     prohibit failure detection by the clustering software. Timeouts used by
+     <command>multipathd</command> and <command>sbd</command> must be properly
+     coordinated to avoid premature fencing.
```
I think multipath timeouts should be lower than fencing timeouts. But again I'm not a true expert  on these matteres.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -390,51 +511,46 @@
+    is redirected to the remaining paths. If all active paths fail,
     inactive secondary paths must be woken up, so failover occurs with a delay
-    of approximately 30 seconds.
+    of up to 30 seconds, depending on the properties of the storage array.
```
This is a property. In particular older array make need a long time for failover.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -453,149 +569,17 @@
    </para>
    <table>
-    <title>Tools in the multipath-tools package</title>
+    <title>Tools in the multipath-tools and kpartx packages</title>
```
Not sure what you mean here

mwilck replied:
ah, you meant table →  list, right? Leaving this open

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -646,9 +633,9 @@
-         Maps linear devmaps to partitions on the multipathed device, which
-         makes it possible to create multipath monitoring for partitions on the
-         device.
+         Command line tool for managing &ldquo;partitions&rdquo; on multipath
```
Because these devices aren't the same thing as SCSI disk partitions, for example. Those are first-class kernel devices, as the kernel has an internal concept of disks and partitions. The kernel "knows" that a certain partitions belongs to a certain disk. That concept doesn't apply to DM devices. kpartx "partitions" are just linear mappings on top of other DM devices. The kernel does not regard them as partitions of their parent device. They look like partitions to user space, and they map to exactly the same sectors as the partitions of the SCSI partitions of the underlying devices, but technically they are different.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -671,389 +658,427 @@
+    </para>
+    <para>
+      <command>multipathd</command> serves the following tasks (some of them
+      depend on the configuration):
```
This is just a summary of what multipathd can do. We'll explain the details later.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -671,389 +658,427 @@
+	<listitem>
+	  <para>
+	    On startup, detects path devices and sets up multipath maps from
+	    detected devices, like the <command>multipath</command> command.
```
ok

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -671,389 +658,427 @@
+	</listitem>
+	<listitem>
+	  <para>
+	    Monitors uevents and device mapper events, adding or removing path
```
multipathd does not add or remove path devices. This is done by the kernel. multipathd just reacts. But it removes a "path" (iow, mapping) from a multipath map when the corresponding path **device** has been deleted.

Suggestions for better wording appreciated.

mwilck replied:
replaced "path devices" by "path mappings". Hope this is better,

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -671,389 +658,427 @@
+	<listitem>
+	  <para>
+	    When all paths fail, either fail the map, or switch the map
+	    device to queuing mode for a given time interval.
```
No this is not "a" queueing mode. There is only one queueing mode. It means that rather than actually submitting I/O to devices, the I/O requests are just queued up  in memory.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -671,389 +658,427 @@
+	</listitem>
+	<listitem>
+	  <para>
+	    Handle SCSI persistent reservation keys for path devices if configured. See
```
Both :smile:  it will be explained in details later

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -671,389 +658,427 @@
+	  notice configuration changes and react to them. In some situations,
+	  the daemon might even try to undo the just-applied changes. Therefore
+	  <command>multipath</command>
+	  automatically delegates certain possibly dangerous commands, like
```
multipath detects a running daemon, and "translates" the multipath command (e.g. "flush map") into a multipathd command which it then sends to the multipathd socket.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -671,389 +658,427 @@
       <para>
-       Selectively suppresses a multipath map, and its device-mapped
-       partitions.
+	Shows the currently known path devices using a format string. Use
```
yes

mwilck replied:
Do you want a different wording here?

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -671,389 +658,427 @@
+	Reread configuration files, rescan devices, and set up maps
+	again. This is basically equivalent to a restart of
+	<command>multipathd</command>. A few options can't be modified
+	without a restart, they are mentioned in
```
Ooops, I meant the man page.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -671,389 +658,427 @@
+  <sect2 xml:id="sec-multipath-mpiotools-multipath">
+   <title>The multipath command</title>
+   <para>
+    <command>multipath</command> was the primary tool to configure and
```
ok. This is one of my problems, a lot of the material in this chapter refers to very old releases :grin:

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -671,389 +658,427 @@
+    </varlistentry>
+  </variablelist>
+  <para>
+    The option <option>-v</option> controls the verbosity of the output. The
```
Where did you take this from? We use max verbosity of 4.

mwilck replied:
I rewrote this paragraph.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -1131,393 +1183,300 @@
-    zoning setup of the SAN. In particular, check whether LUN masking is active
+    If multipath devices are detected and
+    <filename>multipathd.service</filename> is enabled, multipath maps should
+    be created automatically. If this doesn't happen, use commands like
```
I don't understand the question. You mean except `lsscsi`? Yes. It's impossible to describe every possible command here.

lvicoun replied:
I was referring to the wording: use commands like. We can provide just one command, but then I suggest to reword the sentence: If this doesn't happen, use the command <command>lsscsi</command>...

mwilck replied:
If that's a requirement, I'd rather delete the entire paragraph. As noted before, there's a large list of possible debugging / troubleshooting tools. I can't possibly list them all. I can try to brainstorm some more and figure out a reasonable set of commands. But tbh, the entire paragraph addresses people with advanced knowledge. Knowing the name of the command won't be helpful unless they also have some clue understanding the output of the command.

mwilck commented:
@lvicoun , thanks a lot for this extensive review. I have commited most of your suggestions. I'll merge this into my local version now, make some additional changes, and re-push.

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -37,201 +35,246 @@
+	</term>
+	<listitem>
+	  <para>
+	    A subsystem in the Linux kernel for creating virtual block
```
ok, will do

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -284,51 +321,144 @@
+	</listitem>
+      </orderedlist>
+      <para>
+	It is possible to migrate between these three setups, but only for
```
removed

mwilck commented:
@lvicoun , I hope I have addressed the bulk of your issues now. I had to fixup some modifications to make it validate, and added some additional work I did a while ago.

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1492,8 +1465,80 @@ section {
+	     <para>
+	       I/O on the multipath map is queued in the device-mapper
+	       layer and sent to the device when path devices become available
+	       again. This is the safest option to avoid loosing data, but it
```
```suggestion
	       again. This is the safest option to avoid losing data, but it
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1492,8 +1465,80 @@ section {
+	       again. This is the safest option to avoid loosing data, but it
+	       can have negative effects if the path devices don't get
+	       reinstated for a long time. Processes reading from the device
+	       the device will hang in uninterruptible sleep
```
```suggestion
	        will hang in uninterruptible sleep
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1492,8 +1465,80 @@ section {
+	   <term><replaceable>N</replaceable></term>
+	   <listitem>
+	     <para>
+	       <replaceable>N</replaceable> is a positive integer number.
```
```suggestion
	       <replaceable>N</replaceable> is a positive integer.
```

lvicoun commented on xml/storage_multipath.xml:
```diff
@@ -1492,8 +1465,80 @@ section {
+	     <para>
+	       <replaceable>N</replaceable> is a positive integer number.
+	       Keep the map device in queuing mode for <replaceable>N</replaceable>
+	       polling intervals. When the time has elapsed, fail the map
```
```suggestion
	       polling intervals. When the time elapses, <command>multipathd</command> fails the map
```

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -37,201 +35,233 @@
+	</listitem>
+      </varlistentry>
+      <varlistentry>
+	<term>
+	  multipath map
+	</term>
+	<term>
+	  multipath device
+	</term>
```
see above

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -37,201 +35,233 @@
+	</listitem>
+      </varlistentry>
+      <varlistentry>
+	<term>
+	  path device
+	</term>
+	<term>
+	  low-level device
+	</term>
```
see above

mwilck commented on xml/storage_multipath.xml:
```diff
@@ -37,201 +35,233 @@
+	</listitem>
+      </varlistentry>
+      <varlistentry>
+	<term>
+	  WWID
+	</term>
+	<term>
+	  UID
+	</term>
+	<term>
+	  UUID
+	</term>
```
see above

dariavladykina commented on xml/storage_multipath.xml:
```diff
@@ -37,201 +35,233 @@
+	</listitem>
+      </varlistentry>
+      <varlistentry>
+	<term>
+	  host
+	</term>
+	<term>
+	  host system
+	</term>
```
There are very conspicuous extraneous spaces before commas (due to the style sheet formatting). To get rid of these, the simplest way is just to not use <term> tags for each term but to group them.

mwilck replied:
I see, thanks. I'll try to merge this. I hope it won't break merging with #1302.

mwilck commented:
@lvicoun, @dariavladykina: I would strongly prefer not to apply any more corrections in this PR at this point. The set of changes is so large that working on your improvement suggestions is extremely sluggish in the Github UI. Also, as you know I've opened #1302,  and any changes made here will make it more difficult to merge this PR with #1302.

I am responsible for the content and overall structuring of the multipath chapter. Wrt to nuances of formulations and English language, I am doing my best, but I have absolutely no issues in taking suggestions for improvement. The problem with the small suggestions is that at the current stage, they slow down my large-scale work.

I'd be perfectly fine with you making this kind of minor modificationis after my PRs have been merged. It's not necessary that you ask my consent for every single change.
